File: 프로그래머스/lv2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.py
def convertor(time):
    hour,minute = time.split(':')
    return 60*int(hour)+int(minute)

def removeShop(s):
    ns = s.replace("C#",'c').replace("D#",'d').replace('F#','f').replace('G#','g').replace('A#','a')
    return ns

def solution(m, musicinfos):
    answer = []
    m = removeShop(m)
    for i,x in enumerate(musicinfos):
        start,end,name,music = x.split(',')
        music = removeShop(music)
        playTime = convertor(end) - convertor(start)
        origin = music*(playTime // L)+music[:(playTime % L)] if playTime > (L:=len(music)) else music[:playTime]
        if m in origin:
            answer.append((-playTime,i,name))
    return sorted(answer)[0][2] if answer else "(None)"

File: 프로그래머스/lv2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.java
import java.util.*;

class Solution {
    public String solution(String m, String[] musicinfos) {
        String answer = "(None)";
        int pt = 0;
        String heard = replaceShop(m);
        for (String music: musicinfos){
            String[] info = music.split(",");
            int playTime = convertor(info[1]) - convertor(info[0]);
            String name = info[2];
            String part = replaceShop(info[3]);
            String origin = part.repeat(playTime / part.length()) + part.substring(0,playTime % part.length());
            if (origin.contains(heard) && (answer.equals("(None)") || pt < playTime)){
                answer = info[2];
                pt = playTime;
            }
        }
        return answer;
    }
    
    public int convertor(String time){
        return Arrays.stream(time.split(":"))
        .mapToInt(Integer::parseInt)
        .reduce((h, m) -> h * 60 + m)
        .orElse(0);
    }
    
    public String replaceShop(String music) {
        return music.replace("C#","c").replace("D#","d").replace("F#","f").replace("G#","g").replace("A#","a");
    }
}